### DISCUSSION TITLE
selectively using statically generated branches

### DISCUSSION THREAD

**Discussion by kevin-m-kent** [ropensci/targets#324] [Help] [answered]
hi - I'm using tar_map to generate a bunch of different variations of data and I'd like to selectively use some of those generated targets in downstream operations. For instance, take a historical data pull (that is currently one target) and another data pull and build a model on the historical bit and test on another dataset. So basically generating a bunch of different targets and branches with tar_map (and even dynamic branching within tar_map) and then do dynamic branching over targets that span a few independent  branches in a following tar_target call. 

 But I'm having trouble figuring out how to reference the targets and create a target vector that can be dynamically mapped over. What I think I'd like to do is use a regex pattern in the pattern = field for tar_target and then map over the targets that fit that criteria. Is that possible or is there an example that might do this? Thank you!

**Comment by kevin-m-kent** [comment]
Something like this for instance - tar_target(some_models, fit_model(starts_with("historical")), pattern = map(starts_with("historical")))

**Reply by kevin-m-kent** [reply]
I was trying to fiddle a bit with some tidy eval stuff by combining tar_objects and tar_read_raw but I figured there was an easier way to do this. for instance:

```
  tar_target(train_data, tar_objects()[str_detect(tar_objects(), "historical")]),
  tar_target(fitted_models, fit_model(train_data), pattern = map(train_data))
  ```
  
  where fit_model uses tar_read_raw on the train_data target strings

**Comment by wlandau** [comment, marked as answer]
If you want to select some dynamic branches just for testing purposes initially, then the alternative patterns at https://books.ropensci.org/targets/dynamic.html#pattern-construction like `head()` or `sample()` might help. For static branching, you can first generate some target objects with `tar_map()` or `tar_eval()`, check the names, and then mention them in downstream commands. Another round of `tar_eval()` or `tar_map()` can select the targets you need.

``` r
library(targets)
tar_script({
  library(tarchetypes)
  first_targets <- tar_map(tar_target(first, f(x)), values = list(x = letters[1:5]))
  selected_first_targets <- rlang::syms(c("first_a", "first_d"))
  downstream_targets <- tar_map(
    tar_target(downstream, g(x)),
    values = list(x = selected_first_targets)
  )
  list(first_targets, downstream_targets)
})
tar_glimpse()
```

![](https://i.imgur.com/UcRxWJ8.png)

<sup>Created on 2021-02-25 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Reply by kevin-m-kent** [reply]
Ah thank you so much! This is exactly what I was trying to do and it worked in my code. 

Ended up doing this to select targets from a previous tar_map sequence. Seems like tar_objects() no longer has the names argument w/ tidyselect functionality?

```
historical_targets <- syms(tar_objects()[str_detect(tar_objects(), "historical")])
```
edit: sorry, I just saw that I think I have an outdated version. Thanks again.

**Reply by wlandau** [reply]
Glad it works for you. `tar_objects()` just lists the contents of `_targets/objects/`, so it is only usable if you already ran the pipeline.
